Added margin property to FormField

joybhallaa opened the pull request:
<!--- Provide a general summary of the PR in the Title above -->

#### Where should the reviewer start?

#### What testing has been done on this PR?

#### How should this be manually tested?
#### Any background context you want to provide?
Issue #3109 

I have tried to address the issue raise by @ericsoderberghp in #3109 , I have added right,top,left margin variables and then they will be passed in the components.
#### Screenshots (if appropriate)
#### Do the grommet docs need to be updated?
No
#### Should this PR be mentioned in the release notes?
Not sure
#### Is this change backwards compatible or is it a breaking change?
Not sure

oorestisime commented on src/js/components/FormField/FormField.js:
```diff
@@ -171,14 +171,29 @@ class FormFieldContent extends Component {
             theme.global.borderSize[border.size],
           )}px`;
+          marginTop = `-${parseMetricToNum(
+            theme.global.borderSize[border.size],
```
maybe we can avoid doing this calculation 4 times

joybhallaa replied:
Okay, will fix this.

joybhallaa replied:
@oorestisime something in line of 
```
static defaultProps = {
 margin: {
  top: 'top',
  bottom: 'bottom',
  right: 'right',
  left: 'left',
 }
}
```

will work ?

joybhallaa commented:
Sure, will work on it in the upcoming days.

joybhallaa commented:
Hey @oorestisime , so as you said I made margin a prop, and extended it's utility in *base.js*, by doing 

``` 
margin : {
top: 'small',
bottom: 'small,
left: 'small,
right: 'small',
}
```

I also added *margin* to the props,  and then rendering it in the FormFieldBox component, I have also added a test, and updated the snapshot. 
Please see the changes from the last commit and let me know if it's right or not, and give feedback.

ShimiSun commented on src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap:
```diff
@@ -277,7 +270,6 @@ exports[`Form with field 1`] = `
   box-sizing: border-box;
   outline: none;
   max-width: 100%;
-  margin-bottom: 12px;
```
do we know why this snapshot changed? it looks like a potential ui regression.

joybhallaa replied:
As far as I know, when I added it **margin** as a prop, and tried a test, it showed this change. This was before the update of the snapshot, I can try to find what's causing this change. 
Thanks for the review and feedback @ShimiSun!

joybhallaa replied:
@ShimiSun , if I remove the *margin* from the **props** object and pass **{...formField.margin}** instead, *margin-top,margin-bottom,margin-right,margin-left* are being added to my snapshot when running **yarn test**.

![Screenshot from 2019-05-31 14-33-27](https://user-images.githubusercontent.com/35984969/58694624-136ef680-83b1-11e9-9679-d6c9c545ec06.png)

ShimiSun commented on src/js/components/FormField/__tests__/FormField-test.js:
```diff
@@ -64,6 +64,16 @@ test('renders htmlFor', () => {
   expect(tree).toMatchSnapshot();
 });
 
+test('renders margin', () => {
```
Thanks for adding a test!

joybhallaa replied:
Please review the latest commit.

joybhallaa commented:
@ShimiSun @oorestisime I'll really like your input on the new commit.

oorestisime commented:
Hey! I am currently without PC until thursday! Will be the first thing i ll
look! Thanks for working on this!

On Sun, Jun 2, 2019, 07:30 Joy Bhalla <notifications@github.com> wrote:

> @ShimiSun <https://github.com/ShimiSun> @oorestisime
> <https://github.com/oorestisime> I'll really like your input on the new
> commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/grommet/grommet/pull/3113?email_source=notifications&email_token=AA5SAFGNSJRE5MS52ESVI5TPYNEGTA5CNFSM4HPDQZ62YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGODWXNTWY#issuecomment-497998299>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA5SAFAVS5C7CZHJTW2NOITPYNEGTANCNFSM4HPDQZ6Q>
> .
>

oorestisime commented:
Thanks once again for working on this. will take a closer look tonight. A few thoughts:

* right now you added values in the theme. that means all the formfields will have by default a margin at top left and right which is a breaking change (not sure intended)
* might have missed this but are we able to set margin via the props? it seems it is only available from the theme no?
* would like a bit more tests with different margin values to be sure we don't break anything now or later.

if you want we can pair on this when you have some time, on slack or anywhere you like!

ericsoderberghp commented on src/js/components/Form/__tests__/__snapshots__/Form-test.js.snap:
```diff
@@ -76,7 +76,10 @@ exports[`Form update 1`] = `
   box-sizing: border-box;
   outline: none;
   max-width: 100%;
+  margin-top: 12px;
```
This looks like a breaking change.

ericsoderberghp commented on src/js/components/FormField/FormField.js:
```diff
@@ -169,16 +168,8 @@ class FormFieldContent extends Component {
         border.position === 'outer' &&
         (border.side === 'all' || border.side === 'horizontal' || !border.side);
       if (abut) {
-        // marginBottom is set to overlap adjacent fields
```
This is also a breaking change. We need to handle the case where the border is on all sides, and the focus indicator is visible appropriately, when there is no margin between the fields.

oorestisime replied:
so this change needs to get reverted @joybhallaa

ericsoderberghp commented on src/js/themes/base.js:
```diff
@@ -524,7 +524,12 @@ export const generate = (baseSpacing = 24, scale = 6) => {
         margin: { vertical: 'xsmall', horizontal: 'small' },
       },
-      margin: { bottom: 'small' },
+      margin: {
```
This is a breaking change. Why do we need to change the default margin in the base theme?

ericsoderberghp commented:
I agree with @oorestisime that we should support a margin in the theme as well as in the props.

joybhallaa commented:
> Thanks once again for working on this. will take a closer look tonight. A few thoughts:
> 
> * right now you added values in the theme. that means all the formfields will have by default a margin at top left and right which is a breaking change (not sure intended)
> * might have missed this but are we able to set margin via the props? it seems it is only available from the theme no?
> * would like a bit more tests with different margin values to be sure we don't break anything now or later.
> 
> if you want we can pair on this when you have some time, on slack or anywhere you like!

Sure @oorestisime , I would love to pair with you on this, I will go through the code once again. Let me know when you are free? Is it okay to send you a personal message on slack ?

joybhallaa commented:
@ericsoderberghp  thanks for the review, I will work on this with @oorestisime , and will try to get this implemented as soon as possible !

oorestisime commented:
sure thing! ping me there!

oorestisime commented on src/js/components/FormField/FormField.js:
```diff
@@ -193,7 +185,7 @@ class FormFieldContent extends Component {
             : undefined
         }
-        margin={abut ? undefined : { ...formField.margin }}
+        margin={abut ? margin : { ...formField.margin }}
```
this doesn't sound right to me. i think we should apply margin if not abut. something like `margin={abut ? undefined : margin || { ...formField.margin }}` so that we apply the margin property if user wants, otherwise we keep formField.margin for backwards compatibility.

joybhallaa commented:
I will not be able to work on this feature, my laptop has broken down unexpectedly, and I don't know when it will get fixed.

buoyantair commented:
@joybhallaa Ah, sorry that happened, I know exactly how it feels. It's fine if you aren't able to do it now, someone else can pick it up :D Magic ✨ of open source ^^

oorestisime commented:
hey @joybhallaa how is your pc situation going? we can wait a bit more for this if you wish to finish it :) take care!

joybhallaa commented:
I'm sorry I can't help you guys anymore, should I close this PR?

oorestisime commented:
No I will take over! thanks!

oorestisime commented:
@ShimiSun @ericsoderberghp 
SO recap on what i did

- rebased to current master
- refactor margin prop into separate function in generic props so i could use it in formfield
- updated TS
- added a margin prop that precedes all options. If user passes a margin prop no matter what we will render with it.
- if abut and no margin prop i kept current state of things to avoid breaking changes
- if not abut and no margin prop we use theme value.
- added a few tests around this. 

I need a few more eyes on this, some edge cases i had to cover :)

ShimiSun commented:
Thank you for collaborating on this @joybhallaa & @oorestisime! another great contribution 💟
